gui/default/syncthing/core/syncthingController.js

```diff
@@ -2591,17 +2591,21 @@ angular.module('syncthing.core')
 
                     var dataReceived = $http.get(urlbase + '/folder/versions?folder=' + encodeURIComponent($scope.restoreVersions.folder))
                         .success(function (data) {
-                            $.each(data, function (key, values) {
-                                $.each(values, function (idx, value) {
-                                    value.modTime = new Date(value.modTime);
-                                    value.versionTime = new Date(value.versionTime);
-                                });
-                                values.sort(function (a, b) {
-                                    return b.versionTime - a.versionTime;
+                            if (isEmptyObject(data)) {
+                                $scope.restoreVersions.noversions = true;
+                            } else {
```

Comment:
Could you initially set it to `$scope.restoreVersions.versions = null` and then `if !closed && !isEmptyObject { iterate and assign to ...versions`? Then in all the conditions you can check for `!restoreVersions.versions`. No new fields or weird stringify checks to identify an empty object.

Reply:
I've done that now (and actually did try doing something similar a few times before), but there's one problem that I'm unable to solve. Basically, with the current code, the Restore Versions modal loads fine on the first try. However, if you close and re-open it, the following error pops up.

```
TypeError: Cannot read properties of undefined (reading 'expanderClosed')
    at setIcon (jquery.fancytree-all-deps.js:11752:14)
    at Fancytree.nodeRenderStatus (jquery.fancytree-all-deps.js:11848:5)
    at Fancytree.nodeRenderStatus (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:6521:10)
    at Fancytree._super (jquery.fancytree-all-deps.js:1685:22)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:13185:15)
    at Fancytree.nodeRenderTitle (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:13123:11)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:1701:22)
    at Fancytree.nodeRender (jquery.fancytree-all-deps.js:13144:11)
```

I've only managed to get around it by changing the code from

```
        <table id="restoreTree">
          <tbody>
            <tr>
              <td>
                <span ng-if="restoreVersions.versions"translate>Loading data...</span>
                <span ng-if="!restoreVersions.versions" translate>There are no file versions to restore.</span>
              </td>
              <td></td>
            </tr>
          </tbody>
        </table>
```

to something more like

```
        <table id="restoreTree" ng-if="restoreVersions.versions">
          <tbody>
            <tr>
              <td translate>Loading data...</td>
              <td></td>
            </tr>
          </tbody>
        </table>
        <table id="restoreTree" ng-if="!restoreVersions.versions">
          <tbody>
            <tr>
              <td translate>There are no file versions to restore.</td>
              <td></td>
            </tr>
          </tbody>
        </table>
```

which does get rid of the error, but the table code ends up being duplicated. The two messages themselves need to be inside the table cells in order to match the same style (padding, etc.) as when there are actual versioned files rendered there.

Reply:
Can you try readding the condition in L3:
https://github.com/syncthing/syncthing/pull/8539/commits/39249a02cac5c0fabb4b14190007f5bd7a9e56f5#diff-35607bd09efb510801fed738df57073a32758c8f16b7f40693fe007d92fd306dL3  
Not that I'd understand why it helps, but it looks like an unrelated change and with this UI code any change in code may trigger any change in behaviour (maybe not quite as bad, but almost :P ).

Reply:
Yeah, it works with that too (as it's essentially the same as when moved down to the table element), but the problem is that I wanted to move the span element

https://github.com/syncthing/syncthing/pull/8539/files#diff-35607bd09efb510801fed738df57073a32758c8f16b7f40693fe007d92fd306dL3

from outside into the table, so that its position and padding matches the table cells, which prevents the text from jumping that takes place now while the table is being loaded 😕. That's why I was experimenting with two different conditions — one for `restoreVersions.versions` itself, and one for when `restoreVersions.versions` is empty.